edb/pgsql/delta.py

```diff
@@ -281,6 +283,172 @@ class AlterObjectProperty(MetaCommand, adapts=sd.AlterObjectProperty):
+                        (SELECT NULLIF(
+                            (SELECT
+                                version::text
+                            FROM
+                                edgedb."_SchemaSchemaVersion"
+                            FOR UPDATE),
+                            {ql(str(expected_ver))}
+                        )),
+                        'serialization_failure',
+                        msg => (
+                            'Cannot serialize DDL: '
+                            || (SELECT version::text FROM
+                                edgedb."_SchemaSchemaVersion")
+                        )
+                    )
+                INTO _dummy_text
+            '''
+        )
+        self.pgops.add(check)
+        return schema
+
+
+class GlobalSchemaVersionCommand(ObjectMetaCommand):
+    pass
+
+
+class CreateGlobalSchemaVersion(
+    ObjectMetaCommand,
+    adapts=s_ver.CreateGlobalSchemaVersion,
+):
+
+    def apply(
+        self,
+        schema: s_schema.Schema,
+        context: sd.CommandContext,
+    ) -> s_schema.Schema:
+        schema = s_ver.CreateGlobalSchemaVersion.apply(self, schema, context)
+        schema = ObjectMetaCommand.apply(self, schema, context)
+        ver_id = str(self.scls.id)
+        ver_name = str(self.scls.get_name(schema))
+        self.pgops.add(
+            dbops.UpdateMetadataSection(
+                dbops.Database(name=edbdef.EDGEDB_TEMPLATE_DB),
+                section='GlobalSchemaVersion',
+                metadata={
+                    ver_id: {
+                        'id': ver_id,
+                        'name': ver_name,
+                        'version': str(self.scls.get_version(schema)),
+                        'builtin': self.scls.get_builtin(schema),
+                        'internal': self.scls.get_internal(schema),
+                    }
+                }
+            )
+        )
+
+        return schema
+
+
+class AlterGlobalSchemaVersion(
+    ObjectMetaCommand,
+    adapts=s_ver.AlterGlobalSchemaVersion,
+):
+
+    def apply(
+        self,
+        schema: s_schema.Schema,
+        context: sd.CommandContext,
+    ) -> s_schema.Schema:
+        schema = s_ver.AlterGlobalSchemaVersion.apply(self, schema, context)
+        schema = ObjectMetaCommand.apply(self, schema, context)
+        ver_id = str(self.scls.id)
+        ver_name = str(self.scls.get_name(schema))
+
+        expected_ver = self.get_orig_attribute_value('version')
+        lock = dbops.Query(
+            f'''
+                SELECT
+                    description
+                FROM
+                    pg_catalog.pg_shdescription
+                WHERE
+                    objoid = (
+                        SELECT oid
+                        FROM pg_database
+                        WHERE datname = {ql(edbdef.EDGEDB_TEMPLATE_DB)}
+                    )
+                    AND classoid = 'pg_database'::regclass::oid
+                FOR UPDATE
```

Comment:
Looks like AWS RDS doesn't allow locking the `pg_shdescription` catalog (but it does allow `COMMENT` on them):

```sql
postgres=> SELECT * FROM pg_shdescription FOR UPDATE;
ERROR:  permission denied for table pg_shdescription
```

I couldn't find an AWS reference for this limit. This is blocking EdgeQL queries like `ALTER ROLE ... SET password := ...`, and causing some relevant tests to fail. Would it be a problem if we don't lock it pls?

Reply:
Sigh.  We use the lock to prevent concurrent modification of the global schema.  We cannot simply remove the lock, need to find another way.